envoy/upstream/cluster_manager.h

```diff
@@ -73,6 +73,76 @@ class ClusterUpdateCallbacksHandle {
 
 using ClusterUpdateCallbacksHandlePtr = std::unique_ptr<ClusterUpdateCallbacksHandle>;
 
+/**
+ * Status enum for the result of an attempted cluster discovery.
+ */
+enum class ClusterDiscoveryStatus {
+  /**
+   * The discovery process timed out. This means that we haven't yet received any reply from
+   * on-demand CDS about it.
+   */
+  Timeout,
+  /**
+   * The discovery process has concluded and on-demand CDS has no such cluster.
```

Comment:
Later on this seems to also cover removed clusters. Can you elaborate on the semantics in different points in the lifecycle?

Reply:
I'm not sure I understand the question, but I suppose that by "removed clusters" you mean a case when config server is being asked for some specific cluster, it could reply that there is no such cluster (it would be in `removed_resources` field of the `DeltaDiscoveryResponse`, thus the "removed cluster"). This is a part of the [recently-amended XDS spec](https://github.com/envoyproxy/envoy/blob/main/docs/root/api-docs/xds_protocol.rst#knowing-when-a-requested-resource-does-not-exist-1). At this point ODCDS will tell cluster manager to notify the discovery manager that such a cluster does not exist, so callbacks for this cluster name can be invoked immediately instead of waiting for a timeout.

Reply:
Ah, right, makes sense. Can you add comments in `OdCdsApiImpl::onConfigUpdate` to this effect?

Reply:
Done. Please see [the relevant commit](https://github.com/envoyproxy/envoy/pull/18723/commits/99e911cdef7f05f0c41985d8cb9b1073966b11e8).